core/client/fs/src/main/java/alluxio/client/file/cache/CacheManager.java

```diff
@@ -12,20 +12,71 @@
 package alluxio.client.file.cache;
 
 import alluxio.conf.AlluxioConfiguration;
+import alluxio.metrics.MetricKey;
+import alluxio.metrics.MetricsSystem;
+
+import com.codahale.metrics.Counter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import java.io.IOException;
+import java.util.Optional;
+
+import javax.annotation.Nullable;
 
 /**
  * Interface for managing cached pages.
  */
 public interface CacheManager extends AutoCloseable  {
+
   /**
-   * @param conf the Alluxio configuration
-   * @return an instance of {@link CacheManager}
+   * Factory class to get or create a CacheManager.
    */
-  static CacheManager create(AlluxioConfiguration conf) throws IOException {
-    // TODO(feng): make cache manager type configurable when we introduce more implementations.
-    return new NoExceptionCacheManager(LocalCacheManager.create(conf));
+  class Factory {
+    private static final Logger LOG = LoggerFactory.getLogger(Factory.class);
+    private static Optional<CacheManager> sCacheManager = null;
+
+    /**
+     * @param conf the Alluxio configuration
+     * @return current CacheManager, creating a new one if it doesn't yet exist
+     */
+    @Nullable
+    public static synchronized CacheManager get(AlluxioConfiguration conf) throws IOException {
+      // TODO(feng): support multiple cache managers
+      if (sCacheManager == null) {
+        try {
+          sCacheManager = Optional.of(create(conf));
+        } catch (IOException e) {
+          sCacheManager = Optional.empty();
```

Comment:
Should we set this to null? Or the idea is if we fail once we do not retry creating the cache manager in the future? If its the latter, update the javadoc to state we only try once.

Reply:
now we always try to create a new filesystem to make the logic simpler to reason